tools/documentation-code-snippets/update-documentation-code-snippets.ts

```diff
@@ -0,0 +1,397 @@
+            // snippet is active
+            snippetsMap[id].active = false;
+          } else {
+            throw new Error(
+              `Encountered ${CODE_BLOCK_SOURCE_START} in line ${
+                index + 1
+              } of ${filename} containing snippet id "${id}" which most recently already had an END marker. Did you forget to open the snippet again?\nFull line:\n${line}`
+            );
+          }
+        } else {
+          throw new Error(
+            `Encountered ${CODE_BLOCK_SOURCE_START} in line ${
+              index + 1
+            } of ${filename} containing snippet id "${id}" which did not had a START marker yet. Did you forget to start the snippet capture for this id?\nFull line:\n${line}`
+          );
+        }
+      }
+    } else {
+      // ORDINARY LINE
+      // add to all active snippets
+      for (const snippet of Object.values(snippetsMap)) {
+        if (snippet.active) snippet.lines.push(line);
+      }
+    }
+  }
+
+  // Ensure all snippets are closed
+  for (const [id, snippet] of Object.entries(snippetsMap)) {
+    if (snippet.active)
+      throw new Error(
+        `The snippet with the id ${id} is not closed properly. Missing an END marker for it in file ${filename}`
+      );
+  }
+
+  // join lines of the snippets and make sure there is no unnecessary leading whitespace
+  const snippets: Snippet[] = Object.entries(snippetsMap).map(
+    ([id, snippet]) => {
+      // calculate common leading whitespace
+      const whiteSpacesToRemove = snippet.lines.reduce((currentMin, line) => {
+        const leadingWhitespace = line.length - line.trimStart().length;
+        return Math.min(leadingWhitespace, currentMin);
+      }, Number.MAX_VALUE);
+
+      return {
+        snippetId: id,
+        content: snippet.lines
+          .map((line) => line.substring(whiteSpacesToRemove))
+          .join("\n"),
+        sourcePath,
+      };
+    }
+  );
+
+  return snippets;
+}
+
+async function main() {
+  console.log(
+    `Looking for #NEXT_CODE_BLOCK_SOURCE comments in all *.mdx files in: ${DOCS_DIRECTORY}\n`
+  );
+
+  const codeBlockTargets = await getAllCodeBlockTargets();
+  const sourcePaths = getAllUniqueSourcePaths(codeBlockTargets);
+  const requestedSnippets: { sourcePath: string; snippetIds: string[] }[] =
+    sourcePaths.map((sourcePath) => ({
+      sourcePath,
+      snippetIds: getSnippedIdsReferencedInSourcePath(
+        codeBlockTargets,
+        sourcePath
+      ),
+    }));
+
+  console.log(
+    `The code block targets found are expecting snippets in the following directories:
+${requestedSnippets
+  .map((s) => `${s.sourcePath} (ids: ${s.snippetIds.join(",")})`)
+  .join("\n")}\n`
+  );
+
+  console.log("Trying to collect the aforementioned snippets..");
+
+  const sourceFilesWithCodeBlock: { sourcePath: string; filename: string }[] =
+    [];
+
+  for (const repoLocalSourcePath of sourcePaths) {
+    const sourcePath = resolveRepoPath(repoLocalSourcePath);
```

Comment:
Why don't we use `git ls-files` here to simplify finding the files under version control

Reply:
This changes `examples/typescript/documentation` to e.g. `/Users/ansgar/projects/hashicorp/terraform-cdk/examples/typescript/documentation`, so I don't know how git ls-files would help with that.
Or do you mean the line with glob()?

Reply:
Yeah I meant getting the list to traverse

Reply:
Ah! Thanks for that tip. I'll switch this as glob() has problems using the gitignore and that includes in e.g. the generated provider bindings to be included 😅